src/components/checkbox/checkbox.js

```diff
@@ -1,29 +1,132 @@
-function listener(event) {
-  const element = event.target;
-  if (element.tagName === 'INPUT' && element.type === 'checkbox' && element.hasAttribute('checked') !== element.checked) {
-    if (element.checked) {
-      element.setAttribute('checked', '');
-    } else {
-      element.removeAttribute('checked');
+import mixin from '../../globals/js/misc/mixin';
+import createComponent from '../../globals/js/mixins/create-component';
+import InitComponentBySearch from '../../globals/js/mixins/init-component-by-search';
+
+const stateChangeTypes = {
+  true: 'true',
+  false: 'false',
+  mixed: 'mixed',
+};
+
+class Checkbox extends mixin(createComponent, InitComponentBySearch) {
+  /**
+   * Checkbox UI.
+   * @extends CreateComponent
+   * @extends InitComponentBySearch
+   * @param {HTMLElement} element The element working as a checkbox UI.
+   */
+
+  constructor(element, options) {
+    super(element, options);
+    this.element.addEventListener('click', () => this._handleClick());
+    this.element.addEventListener('focus', () => this._handleFocus());
+    this.element.addEventListener('blur', () => this._handleBlur());
+
+    this._indeterminateCheckbox();
+    this._initCheckbox();
+  }
+
+  _handleClick() {
+    if (this.element.checked === true) {
+      this.element.setAttribute('checked', '');
+      this.element.setAttribute('aria-checked', 'true');
+      this.element.checked = true;
+
+      // nested checkboxes inside labels
+      if (this.element.parentElement.classList.contains('bx--checkbox-label')) {
+        this.element.parentElement.setAttribute('data-contained-checkbox-state', 'true');
+      }
+    } else if (this.element.checked === false) {
+      this.element.removeAttribute('checked');
+      this.element.setAttribute('aria-checked', 'false');
+      this.element.checked = false;
+
+      // nested checkboxes inside labels
+      if (this.element.parentElement.classList.contains('bx--checkbox-label')) {
+        this.element.parentElement.setAttribute('data-contained-checkbox-state', 'false');
+      }
+    }
+  }
+
+  _handleFocus() {
+    if (this.element.parentElement.classList.contains('bx--checkbox-label')) {
+      this.element.parentElement.classList.add('bx--checkbox-label__focus');
+    }
+  }
+
+  _handleBlur() {
+    if (this.element.parentElement.classList.contains('bx--checkbox-label')) {
+      this.element.parentElement.classList.remove('bx--checkbox-label__focus');
+    }
+  }
+
+  /**
```

Comment:
Let’s remove the duplicated comments here 😉

Reply:
Thank you so much for all your help with this 👍 🎉